src/commands/edge_app.rs

```diff
@@ -0,0 +1,250 @@
+        map.remove("id");
+        map.remove("created_at");
+        map.remove("created_by");
+        map.remove("updated_at");
+        map.remove("permissions");
+
+        let yaml = serde_yaml::to_string(map)?;
+        let input = File::create(path)?;
+        write!(&input, "{yaml}")?;
+
+        Ok(())
+    }
+
+    pub fn publish(self, path: &Path) -> Result<EdgeAppManifest, CommandError> {
+        let url = format!("{}/v4/edge_apps", &self.authentication.config.url);
+
+        let data = fs::read_to_string(path)?;
+        // by converting to struct we make sure there no extra fields and all required fields are present
+        let edge_app: EdgeAppManifest = serde_yaml::from_str(&data)?;
+        let mut binding = serde_json::to_value(edge_app)?;
+        let edge_app_json = binding.as_object_mut().ok_or(CommandError::Serde)?;
+
+        // Id can not be empty when posting. Depending on what we decide I should either raise an error
+        // or we allow users to supply an id.
+        if edge_app_json.contains_key("id")
+            && edge_app_json["id"]
+                .as_str()
+                .ok_or(CommandError::Serde)?
+                .is_empty()
+        {
+            edge_app_json.remove("id");
+        }
+
+        edge_app_json.remove("created_at");
+        edge_app_json.remove("created_by");
+        edge_app_json.remove("updated_at");
+        edge_app_json.remove("permissions");
+
+        let mut payload = HashMap::<String, String>::new();
+        // for now all values are fields in the manifest are required to be non-empty.
+        // if we change that we will need to have a list of required non-empty values.
+        debug!("Edge app headers: ");
+        for (k, v) in edge_app_json {
+            debug!("{k}: {v:?}");
+            let value = v.as_str().ok_or(CommandError::MissingField)?;
+            if value.is_empty() {
+                return Err(CommandError::InvalidManifestValue(k.to_string()));
+            };
+
+            payload.insert(k.to_owned(), value.to_owned());
+        }
+
+        let mut headers = HeaderMap::new();
+        headers.insert("Prefer", "return=representation".parse()?);
+
+        let response = self
+            .authentication
+            .build_client()?
+            .post(url)
+            .headers(headers)
+            .json(&payload)
+            .send()?;
+
+        if response.status() != StatusCode::CREATED {
+            return Err(CommandError::WrongResponseStatus(
+                response.status().as_u16(),
+            ));
+        }
+
+        // I think there is no need to check for size - this vector should always be size 1.
+        // We can let it crash in run time otherwise.
+        let updated_manifests: Vec<EdgeAppManifest> = serde_json::from_str(&response.text()?)?;
+
+        // overwrite manifest file now that we have id and extra fields.
+        let yaml = serde_yaml::to_string(&updated_manifests[0])?;
```

Comment:
Should we also drop created_by, etc fields from manifest?

manifest data is:
```
--  Manifest data
    name         text        NOT NULL,
    version      text        NOT NULL,
    description  text        NOT NULL,
    icon         text        NOT NULL,
    author       text        NOT NULL,
    homepage_url text        NOT NULL,
    permissions  text[]      DEFAULT '{}'
```

and id.

Reply:
Maybe - is that data even relevant for users all the time? can be somewhere in another table.

Reply:
It is a property of edge app - no need to separate it, even in the view.
What you could do is select only required fields when doing API request:

?select=name,version,description...

https://postgrest.org/en/stable/api.html#embedding-after-insertions-updates-deletions